File: backend/utils/nasa_api.py
import os
from dotenv import load_dotenv
import requests

load_dotenv()  # loads .env in backend/
NASA_API_KEY = os.getenv("NASA_API_KEY", "DEMO_KEY")
BASE_URL = "https://api.nasa.gov"

def fetch_apod(date=None):
    try:
        params = {"api_key": NASA_API_KEY}
        if date:
            params["date"] = date
        r = requests.get(f"{BASE_URL}/planetary/apod", params=params, timeout=10)
        r.raise_for_status()
        return r.json()
    except Exception as e:
        return {"error": str(e)}

def fetch_mars_photos(sol=1000, camera=None):
    try:
        params = {"sol": sol, "api_key": NASA_API_KEY}
        if camera:
            params["camera"] = camera
        r = requests.get(f"{BASE_URL}/mars-photos/api/v1/rovers/curiosity/photos", params=params, timeout=10)
        r.raise_for_status()
        data = r.json()
        photos = data.get("photos", [])
        if not photos:
            return {"message": "No photos found for this sol/camera. Try other sol values or remove camera param."}
        return photos[:12]  # limit to 12
    except Exception as e:
        return {"error": str(e)}

def fetch_epic(date=None):
    try:
        if date:
            url = f"{BASE_URL}/EPIC/api/natural/date/{date}"
        else:
            url = f"{BASE_URL}/EPIC/api/natural"
        params = {"api_key": NASA_API_KEY}
        r = requests.get(url, params=params, timeout=10)
        r.raise_for_status()
        return r.json()
    except Exception as e:
        return {"error": str(e)}

def fetch_neo(start_date=None, end_date=None):
    try:
        params = {"api_key": NASA_API_KEY}
        if start_date:
            params["start_date"] = start_date
        if end_date:
            params["end_date"] = end_date
        r = requests.get(f"{BASE_URL}/neo/rest/v1/feed", params=params, timeout=10)
        r.raise_for_status()
        return r.json()
    except Exception as e:
        return {"error": str(e)}

File: src/lib/nasaApi.js
import axios from "axios";

const NASA_API_KEY = import.meta.env.VITE_NASA_API_KEY || "DEMO_KEY";
const BASE_URL = "https://api.nasa.gov";

const nasaAxios = axios.create({
  baseURL: BASE_URL,
  timeout: 15000,
});

nasaAxios.interceptors.response.use(
  (response) => response,
  (error) => {
    const fallback =
      error.response?.data?.error ||
      error.response?.data?.message ||
      error.message ||
      "Unexpected error";
    return Promise.reject(new Error(fallback));
  }
);

export const fetchApod = async (date = null) => {
  try {
    const params = { api_key: NASA_API_KEY };
    if (date) {
      params.date = date;
    }
    const response = await nasaAxios.get("/planetary/apod", { params });
    return response.data;
  } catch (error) {
    return { error: error.message };
  }
};

export const fetchMarsPhotos = async (sol = 1000, camera = null) => {
  try {
    const params = { sol, api_key: NASA_API_KEY };
    if (camera) {
      params.camera = camera;
    }
    const response = await nasaAxios.get(
      "/mars-photos/api/v1/rovers/curiosity/photos",
      { params }
    );
    const photos = response.data.photos || [];
    if (!photos.length) {
      return {
        message:
          "No photos found for this sol/camera. Try other sol values or remove camera param.",
      };
    }
    return photos.slice(0, 12);
  } catch (error) {
    return { error: error.message };
  }
};

export const fetchEpic = async (date = null) => {
  try {
    let url = "/EPIC/api/natural";
    if (date) {
      url = `/EPIC/api/natural/date/${date}`;
    }
    const params = { api_key: NASA_API_KEY };
    const response = await nasaAxios.get(url, { params });
    return response.data;
  } catch (error) {
    return { error: error.message };
  }
};

export const fetchNeo = async (startDate = null, endDate = null) => {
  try {
    const params = { api_key: NASA_API_KEY };
    if (startDate) {
      params.start_date = startDate;
    }
    if (endDate) {
      params.end_date = endDate;
    }
    const response = await nasaAxios.get("/neo/rest/v1/feed", { params });
    return response.data;
  } catch (error) {
    return { error: error.message };
  }
};
